content/Software and Downloads/Compilation/If-your-sketch-doesnt-compile.md

```diff
@@ -0,0 +1,242 @@
+---
+title: "If your sketch doesn't compile"
+---
+
+Learn about compilation errors in your sketch and how to resolve them.
+
+---
+
+## Quick checks
+
+* Make sure your error occurs during compilation by clicking ![Verify button](img/symbol_verify.png) **Verify** instead of ![Upload button](img/symbol_upload.png) **Upload**. This will compile the sketch without attempting to upload it. If your error only occurs when uploading, see [Errors when uploading a sketch](https://support.arduino.cc/hc/en-us/articles/4403365313810-Errors-when-uploading-a-sketch).
+
+* A successful compilation will always finish with this message:
+
+  ```
+  Sketch uses 11604 bytes (4%) of program storage space. Maximum is 262144 bytes.
+  Global variables use 2980 bytes (9%) of dynamic memory, leaving 29788 bytes for local variables. Maximum is 32768 bytes.
+  ```
+
+* A sketch always needs to include a `setup()` and `loop()` function, even if they're not being used. You can use _File > Examples > 1.Basics > BareMinium_ as a template.
+
+* Libraries added with `#include` need to be installed. Try searching for the library name in the Library Manager (_Tools > Manage Libraries..._).
+
+---
+
+## Interpreting the error message
+
+Arduino IDE displays compilation messages differently depending on the version:
+
+1. Check the IDE error message.
+   * In IDE 1.x, it is displayed above the console window.
+   * In IDE 2.x, it is displayed in a bottom-right pop-up.
+   Sometimes, preceding lines can be more informative — check the console output.
+
+2. Check the console output.
+   * Compilation errors will often reference the row and column where the error was triggered, i.g. `Blink:29:1` (line 29, 1 column).
+   * You may have to scroll or expand the window width to see the entire message.
+
+3. Look for highlights in the editor:
+   * In IDE 1.x, the **line** where the error occurred is highlighted in red.
+   * In IDE 2.x, the **character** where the error occurred is underlined in red.
+   * The line causing the error may be before the line where it was triggered. In the example above, line 28 is missing a semicolon (`;`), but the error will be triggered by the unexpected closing bracket (`}`) in line 29.
+
+   <table class="img-captions">
+     <tr>
+        <td><img src="img/ide1-syntax-error-output-highlighted.png" alt="Compilation error in IDE 1.8.13, due to a missing semicolon."></td>
+        <td><img src="img/ide2-syntax-error-output-highlighted.png" alt="Compilation error in IDE 2.0.0-3c6, due to a missing semicolon."></td>
+     </tr>
+     <tr>
+        <td>Error output in IDE 1.x.</td>
+        <td>Error output in IDE 2.x.</td>
+     </tr>
+   </table>
+
+---
+
+## Troubleshoot a specific error
+
+<table>
```

Comment:
Missing links to anchors in some of the errors

Reply:
6134b5e